### PR TITLE
Issue #9146: switch to jdk 11

### PR DESCRIPTION
**Pull request by pbludov**
Issue #9146

### First commit
* explicit release version for validation.sh javac8
* javafx-tests-for-jdk11 profile of equalsverifier disabled
* docker image jdk-8-groovy-git-mvn -> jdk-11-groovy-git-mvn for circleci
* docker image maven:3.6.3-adoptopenjdk-8 -> maven:3.8.4-adoptopenjdk-11 for drone
*  java-version: 8 ->  java-version: 11 for github workflows
* maven 3.8.4 for appveyor, JDK8 builds removed
* JDK13, JDK14 builds in azure pipelines removed, JDK17 added
* Set minimal required version to Java 11

### Second commit
* Move all Java{9-11} inputs to compilable resources folder

### Third commit
* Fixes for IDEA inspections triggered for Java 11 build

### CODE REVIEW

File: .ci/validation.sh
Comment (pbludov) [inline .ci/validation.sh:715]: Equalsverifier has a profile for Java 11+:
https://github.com/jqno/equalsverifier/blob/main/pom.xml#L352
```xml
        <profile>
            <id>enables-javafx-tests-for-jdk11-and-up</id>
            <activation>
                <jdk>[11,)</jdk>
            </activation>

            <dependencies>
                <dependency>
                    <groupId>org.openjfx</groupId>
                    <artifactId>javafx-base</artifactId>
                    <version>17.0.1</version>
                    <scope>provided</scope>
                </dependency>
            </dependencies>
```
This profile requires JavaFX. Installing JavaFX for ppc64le linux is not trivial, and it is not actually necessary, so this profile is disabled.

File: azure-pipelines.yml
Comment (pbludov) [inline azure-pipelines.yml:86]: `export JAVA_HOME=$JAVA_HOME_11_X64` does not work with `eval "$(cmd)"`
Replaced as one command with environment variable override.

**Review by rnveach — APPROVED** [review]

**Comment by romani** [comment]
do we have jdk11 Input files that we need to move to compilable folder ?

**Comment by pbludov** [comment]
> do we have jdk11 Input files that we need to move to compilable folder ?

Good point. There should be few of them. I will update the PR.

**Comment by pbludov** [comment]
Java9 and Java10 test sources should be moved too.

**Comment by pbludov** [comment]
@romani sorry, we can't move jdk9,10,11 files to compilable folder without changing `java.version` in the `pom.xml`. The build fails with 
```
[ERROR] src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/matchxpath/InputMatchXpathAvoidInstanceCreationWithoutVar.java:[14,9] cannot find symbol
[ERROR]   symbol:   class var
[ERROR]   location: class com.puppycrawl.tools.checkstyle.checks.coding.matchxpath.InputMatchXpathAvoidInstanceCreationWithoutVar
```
and many other errors like this.

**Review by nmancus1 — APPROVED** [review]

**Comment by romani** [comment]
I propose to make second commit in this PR and start migration. This PR will be merged after 9.3 release.

**Review by strkkk — APPROVED** [review]

**Comment by rnveach** [comment]
@pbludov Does xwiki have Java 11 POM now too?
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/11299/workflows/75ec714d-b7ca-4341-a1a9-a2bbf66058fd/jobs/111858?invite=true&invite=true#step-102-341

**Comment by pbludov** [comment]
> @pbludov Does xwiki have Java 11 POM now too? https://app.circleci.com/pipelines/github/checkstyle/checkstyle/11299/workflows/75ec714d-b7ca-4341-a1a9-a2bbf66058fd/jobs/111858?invite=true&invite=true#step-102-341

Yes, since this commit: https://github.com/xwiki/xwiki-commons/commit/1a919e0252a93c8d1cfdce2730c7ebe58e0697e2

**Comment by romani** [comment]
@pbludov , conflict.
Xwiki migrated to jdk11, merged urgently.

**Comment by pbludov** [comment]
> @pbludov , conflict. Xwiki migrated to jdk11, merged urgently.

PR rebased

**Comment by romani** [comment]
Can we do final update of jdk and input files in second commit ?

**Comment by romani** [comment]
Github, rebase

**Comment by romani** [comment]
I will deal with Team City , let's proceed with review of this PR.

File: .ci/validation.sh
Comment (romani) [inline .ci/validation.sh:496]: Why we need javac8  validation ?

File: .ci/validation.sh
Comment (romani) [reply .ci/validation.sh:727]: Author of this repo told that we need jdk17  for static analysis tools
Do we have jdk17 in CI ?

File: .ci/validation.sh
Comment (romani) [reply .ci/validation.sh:496]: Can we remove https://github.com/checkstyle/checkstyle/blob/3a790e87be85f2adfd75fe9e02f16f00deec2f98/.ci/validation.sh#L499

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: .ci/validation.sh
Comment (pbludov) [reply .ci/validation.sh:496]: > Why we need javac8 validation ?

Some input files are incompatible with Eclipse compiler. These inputs should be compiled with javac11 now.

File: .ci/validation.sh
Comment (pbludov) [reply .ci/validation.sh:727]: Yes, we even have jdk18 in Travis. I'll try to fix Equalsverifier in a separate PR

File: .ci/validation.sh
Comment (pbludov) [reply .ci/validation.sh:496]: done

File: .ci/validation.sh
Comment (pbludov) [reply .ci/validation.sh:713]: > Do we have jdk17 in CI ?

Sorry, latest build for [AdoptOpenJdk](https://adoptopenjdk.net/installation.html) is jdk16.

**Comment by romani** [comment]
>this PR is NOT to solve the issue, but to set up all prerequisites

Is this still actual ?

File: .ci/validation.sh
Comment (romani) [reply .ci/validation.sh:496]: Yes, all should be jdk11, jdk8 should gone completely

**Comment by romani** [comment]
Please squash first two commits in one.

Please share grep over sources to prove that no leftovers are left for jdk8, javac8/9/20

File: .ci/validation.sh
Comment (jqno) [reply .ci/validation.sh:713]: Note that the `enables-javafx-tests-for-jdk11-and-up` profile no longer exists, so you can remove that if you like.

Also, as I mentioned in https://github.com/jqno/equalsverifier/issues/586#issuecomment-1025459469, I can make it so that Checkstyle in the EqualsVerifier build will run on a JDK of your choice, just let me know in that ticket and I'll make it happen.

**Comment by romani** [comment]
@pbludov , please proceed with updates. I am waiting for TC support to respond on jdk switch.

**Comment by pbludov** [comment]
> > this PR is NOT to solve the issue, but to set up all prerequisites
> Is this still actual ?

No. but I'd like make all non-breaking changes in this PR (the first commit only). Then all other stuff.

**Comment by pbludov** [comment]
> Please share grep over sources to prove that no leftovers are left for jdk8, javac8/9/10

Done: https://gist.github.com/pbludov/655e6b4de7aba630063c80018ae62899

**Comment by pbludov** [comment]
> I found problem with TC config.
Confirmed.  IDEA Inspections now fails on some Java9 and Java11 inspections.  I'm going to fix it asap.

**Comment by pbludov** [comment]
IDEA is still failing:
https://teamcity.jetbrains.com/viewLog.html?buildId=3778971&tab=Inspection&buildTypeId=Checkstyle_IdeaInspectionsPullRequest

This is a false positive:
```java
    protected static String readFile(String filename) throws IOException {
        return toLfLineEnding(Files.readString(Paths.get(filename)));
    }
```
The exception *may be* thrown by `Files.readString`.

**Comment by pbludov** [comment]
@romani done

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by rnveach** [comment]
> Ok to merge

Just double confirming since once it is merged, we are breaking compatibility and going to a new major version.

We are ok to merge this for this release?

**Comment by romani** [comment]
Yes, it  means we bump to 10.0 version

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:81]: This is not really appropriate to place into the main Javadoc documentation. Can we not attach this to the `noinspection`?

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:1156]: Same and others....

File: src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:81]: This is how we always do this.
We do not know now to record reasons of suppression

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:81]: TC won't allow us to do:
> `@noinspection RedundantThrows - False positive.`

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:81]: TC won't allow us to do

No. Unfortunately.

>`@tcReason`

It triggers some another  violation. We can investigate it in separate issue one more time.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:81]: "mvn site" is failing:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.10.0:site 
(default-site) on project checkstyle: Error generating maven-javadoc-plugin:3.3.1:javadoc report: 
[ERROR] Exit code: 1 - /home/rivanov/java/github/romani/checkstyle/src/main/java/
com/puppycrawl/tools/checkstyle/Main.java:631: error: unknown tag: noinspectionreason
[ERROR]          * @noinspectionreason: CanBeFinal - we use picocli and it use
[ERROR]            ^
[ERROR] 
[ERROR] Command line was: /opt/jvm/jdk1.8.0_162/jre/../bin/javadoc @options @packages

```

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/AbstractPathTestSupport.java:81]: new issue - https://github.com/checkstyle/checkstyle/issues/11277

we currently update javadoc on Test classes, nobody read this, bigger  problem will be addressed in separate issue.

**Comment by romani** [comment]
rebased to resolve conflict.

**Comment by pbludov** [comment]
> `.ci/validation.sh nondex` should be failing, unless I am missing something.

This is really strange. Other PRs:
image

This PR:
image

One check is missing: Semaphore CI.

**Comment by pbludov** [comment]
> > Also, can we update comments:
> 
> Done.

```bash
$  grep -R "jdk8" | grep yml
```

**Comment by pbludov** [comment]
Blocked by #11277

**Comment by romani** [comment]
Let's not be blocked.

We can improve all in separate PRs.

**Comment by nmancus1** [comment]
Also, failing nondex execution (really failing maven execution in this case) is not failing build: https://checkstyle.semaphoreci.com/jobs/ef492635-0104-483a-a1d5-21507f7084ff

**Comment by pbludov** [comment]
> Semaphore is failing and needs updates https://checkstyle.semaphoreci.com/workflows/77fe5d68-d5ca-4a45-ae4f-61716cb09dd2?pipeline_id=d4794fff-96f9-4d7e-a20b-bd4d05f019e4

This is an old version of this PR. It looks like something is lost on the semaphore side. I'll try to re-create this PR

**Comment by pbludov** [comment]
No luck. #11278 completed without Semaphore.

**Comment by pbludov** [comment]
I'm out of ideas. We can merge it and look at the master CI build. Maybe this is a random Semaphore CI failure.

**Comment by romani** [comment]
We can reexecute all shell command on our local. All Cls jobs are executable on local.

**Comment by pbludov** [comment]
Extracted semaphore.yml changes to a separate PR #11279

**Comment by romani** [comment]
please remove jdk11 specific by very general executions:
https://github.com/pbludov/checkstyle/blob/29a87808ef50ee72b4f71e9ec91b55401cd9dccc/.semaphore/semaphore.yml#L53-L54


stable failure at:
```
.ci/validation.sh all-sevntu-checks
config/checkstyle_sevntu_checks.xml:5.61: failed to load external entity "https://checkstyle.org/dtds/configuration_1_2.dtd"
        "https://checkstyle.org/dtds/configuration_1_2.dtd">

```

**Comment by romani** [comment]
> please remove jdk11 specific by very general executions
no need to pre-build checkstyle:

done

**Review by rnveach — APPROVED** [review]

**Comment by rnveach** [comment]
@romani told me merging would waiting on him to verify his local configuration.

**Comment by romani** [comment]
I am merging, will do direct commits to bump version to 10.0 directly to master, milestone update, ....

**Comment by pbludov** [comment]
> I am merging, will do direct commits to bump version to 10.0 directly to master, milestone update, ....

Semaphore CI is green on master after merging this PR.

**Pull request merged**
